### PR TITLE
[#238] Finish the transition process of splitting the group record and membership list on two separate collections

### LINKED ISSUE #238: [FEATURE] Finish the transition process of splitting the group record and membership list on two separate collections
**Is your feature request related to a problem? Please describe.**
Finish the transition process of splitting the group record and membership list on two separate collections.
The current state fixes only the bigger groups and this is an attempt to bring this transition to the end

**Describe the solution you'd like**
We should migrate all of the old embedded memberships to the new group_memberships collection and update all APIs to look for memberships in this collection.

### PR DESCRIPTION
**Pull request by mdryankov**
## Description

**Resolves #238**

[comment]: # (This project only accepts pull requests related to open issues.)
[comment]: # (If suggesting a new feature or change, please discuss it in an issue first.)
[comment]: # (If fixing a bug, there should be an issue describing it with steps to reproduce.)

## Review Time Estimate
_Please give your idea of how soon this pull request needs to be reviewed by selecting one of the options below. This can be based on the criticality of the issue at hand and/or other relevant factors._

[comment]: # (To select an option, please put an 'x' in the applicable box.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [x] Immediately
- [ ] Within a week
- [ ] When possible

## Type of changes
_Please select a relevant option:_

[comment]: # (To select an option, please put an 'x' in the applicable box.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] Other (any another change that does not fall in one of the above categories.)

## Checklist:
_Please select all applicable options:_

[comment]: # (To select your options, please put an 'x' in the all boxes that apply.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [ ] I have signed the Rokwire Contributor License Agreement ([CLA](https://rokwire.org/rokwire_cla)). (_Any contributor who is not an employee of the University of Illinois whose official duties include contributing to the Rokwire software, or who is not paid by the Rokwire project, needs to sign the CLA before their contribution can be accepted._)
- [x] I have updated the [CHANGELOG](../CHANGELOG.md).
- [ ] I have read the [Contributor Guidelines](../CONTRIBUTING.md).
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] My change requires updating the documentation.
- [ ] I have made necessary changes to the documentation.
- [ ] I have added tests related to my changes.
- [ ] My changes generate no new warnings.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.

---

[comment]: # (Template credit: This pull request template is based on Embedded Artistry {https://github.com/embeddedartistry/templates/blob/master/.github/PULL_REQUEST_TEMPLATE.md}, Clowder {https://github.com/clowder-framework/clowder/blob/develop/.github/PULL_REQUEST_TEMPLATE.md}, and TalAter {https://github.com/TalAter/open-source-templates} templates.)

### CODE REVIEW

File: core/services.go
Comment (shurwit) [inline core/services.go:899]: We should probably rewrite this whole `synchronizeAuthman` function since it is currently putting the admins in as  embedded members.  It would be better if it just directly updated the group memberships as needed or passed the admin data to `synchronizeAuthmanGroup` to be handled there.

File: driven/storage/adapter.go
Comment (shurwit) [inline driven/storage/adapter.go:1123]: This needs to be handled in a transaction to avoid race conditions

File: driven/storage/adapter_v3.go
Comment (shurwit) [inline driven/storage/adapter_v3.go:15]: I don't think that filter should be a pointer since I do not believe we would want to allow this function to be called with a nil filter. If we do want to support this, we would need to handle the nil case first so we don't dereference a nil pointer below.

File: driven/storage/adapter_v3.go
Comment (shurwit) [inline driven/storage/adapter_v3.go:40]: The filter being passed to this function does not contain anything so presumably it will return all group memberships? I assume we would want to pass `&model.MembershipFilter{UserID: filter.MemberUserID}`. Also, do we want to do this if we have already filtered by `MemberID` or `MemberExternalID`? It looks like you are trying to apply only one user related filter. It also seems like we should be setting `userID = filter.MemberUserID` here.

File: driven/storage/adapter_v3.go
Comment (shurwit) [inline driven/storage/adapter_v3.go:119]: Again, I don't think we want to make `filter` a pointer. If we do, we need nil handling

File: core/services.go
Comment (mdryankov) [reply core/services.go:92]: This accidental change detected something more interesting. Those categories seem are not user at all for quite a long time. It looks the client is retrieving event categories and set as a group category and this change is prior introducing the flutter plugin (year or so old).

Comment (mdryankov) [reply core/services.go:92]: I'm not sure who is initiator of this request and why. Basically we could remove the API in this case...

File: driven/storage/adapter.go
Comment (mdryankov) [reply driven/storage/adapter.go]: Nice catch! Thanks!

File: driven/storage/adapter.go
Comment (mdryankov) [reply driven/storage/adapter.go:719]: UpdateGroupWithMembers is planned for the purgatory

File: driven/storage/adapter_v3.go
Comment (mdryankov) [reply driven/storage/adapter_v3.go:16]: It may be invoked in any kind of configuration with and without a filter. Prefer to check for nil. Thanks for the catch

File: driven/storage/adapter_v3.go
Comment (mdryankov) [reply driven/storage/adapter_v3.go]: Nice catch! Thanks!

File: driven/storage/adapter_v3.go
Comment (mdryankov) [reply driven/storage/adapter_v3.go]: Yes. This was one of few items that have been inconsistent. Thanks for those catches!

File: driven/storage/adapter_v3.go
Comment (mdryankov) [reply driven/storage/adapter_v3.go]: Here I agree with you. Here the filter is a mandatory element and at least one filter must be defined. We don't want to respond with all records in the collection

Comment (mdryankov) [reply driven/storage/adapter_v3.go]: The pointer has been removed

File: core/services.go
Comment (shurwit) [reply core/services.go]: Thanks for looking into this. I think we should go ahead and remove it if it is not used/needed

File: core/services.go
Comment (shurwit) [reply core/services.go:895]: Thanks @mdryankov, I think we should make this change before merging this PR since I think our goal should be to have the `members` field remain empty after the migration once this is merged. Let me know if you would like to take a pass at re-writing this, or if you would like me to give it a shot. Thanks again!

File: core/services.go
Comment (mdryankov) [reply core/services.go]: Yes and thanks for the feedback. This is not used by the client and produce a noice. Once we have clarity, then we could redesign and apply in an appropriate way.

File: driven/storage/adapter.go
Comment (mdryankov) [reply driven/storage/adapter.go]: @shurwit, this method has been updated slightly but still available

**Comment by mdryankov** [comment]
@shurwit @roberlander2 , Please review the current state, because today I had to merge a conflict with develop which is related to the react to post functionality. I'm not sure [238-migrate-group-memebrships] contains everything from develop. 
Please add the missing pieces and after that make some intermediate merge, because the difference are getting out of control.

**Comment by roberlander2** [comment]
Hi @mdryankov, it looks to me that this PR has everything from develop (post reactions, get post API, get members by list of user IDs). Thanks!

File: core/services.go
Comment (mdryankov) [reply core/services.go:741]: @shurwit I think the members within the group model are used only for migration purpose. Please resolve this entry if it's fine.

**Comment by mdryankov** [comment]
> Hi @mdryankov, it looks to me that this PR has everything from develop (post reactions, get post API, get members by list of user IDs). Thanks!

@roberlander2  Many thanks for the feedback!

**Comment by mdryankov** [comment]
Hi @shurwit , Today changes include fixes for the reported issues:

- Cannot approve a pending member but request to join works
- Member counts don’t seem correct in some groups
- Cannot change a member to Admin
- Cannot remove a member

Notice the Member filter elements in GroupsFilter model. I'm doubt the MemberID, MemberExternalID, MemberUserID filter is good approach. Feel free to propose another approach which could prevent wrong intentions

File: driver/web/rest/internal.go
Comment (shurwit) [reply driver/web/rest/internal.go:278]: Hi @mdryankov, please take a look at #227 for details on improvements here and let me know if you have any feedback. Thanks!

**Comment by shurwit** [comment]
> Hi @shurwit , Today changes include fixes for the reported issues:
> 
> * Cannot approve a pending member but request to join works
> * Member counts don’t seem correct in some groups
> * Cannot change a member to Admin
> * Cannot remove a member

Hi @mdryankov, thanks for taking care of these issues!

> Notice the Member filter elements in GroupsFilter model. I'm doubt the MemberID, MemberExternalID, MemberUserID filter is good approach. Feel free to propose another approach which could prevent wrong intentions

I think the filter is fine the way it is currently set up since it keeps it clear which ID is being used as you mentioned.

**Comment by mdryankov** [comment]
@shurwit This PR has some readiness to be merged unless we find something missed. Feel free to point such stuff or put changes and if everything is fine we could proceed with a merge next days.

**Comment by shurwit** [comment]
> @shurwit This PR has some readiness to be merged unless we find something missed. Feel free to point such stuff or put changes and if everything is fine we could proceed with a merge next days.
> 
> Also this PR includes the Stats changes for #227

Thanks @mdryankov! I will review this as soon as I get a chance. @roberlander2 will also take a look. Thanks again for all your work on this!

**Review by roberlander2 — CHANGES_REQUESTED** [review]
Thanks @mdryankov, I think this is looking good. I left a few comments below for some minor fixes.

File: driver/web/rest/adminapis.go
Comment (mdryankov) [reply driver/web/rest/adminapis.go]: Fixed

File: core/services_v3.go
Comment (mdryankov) [reply core/services_v3.go]: Changed

File: core/services_v3.go
Comment (mdryankov) [reply core/services_v3.go]: Simplified. Thanks for the catch!

File: driven/storage/adapter_v3.go
Comment (mdryankov) [reply driven/storage/adapter_v3.go]: Quite a good point! I decided to remove it as an overcomplicated block and really hard to read. 
The idea was to cover the privacy permission, but however the intention who requests this method and for what took out of control and need to step back.

Comment (mdryankov) [reply driven/storage/adapter_v3.go]: @shurwit  @roberlander2 Let me know please your thoughts.

File: driven/storage/adapter_v3.go
Comment (mdryankov) [reply driven/storage/adapter_v3.go]: Thanks for this. I would prefer to use groupIDs as inner storage, because the filter is accessible outside of this method and it would be a kind of inconsistency to add more groupIDs. I had some doubts to remove all Member filters, because they produce missconception and easily may be used in a wrong direction. 
@shurwit @roberlander2 Let me know please what do you think?

File: driven/storage/adapter_v3.go
Comment (mdryankov) [reply driven/storage/adapter_v3.go:364]: Agree. Thanks for the good point!

File: driven/storage/adapter_v3.go
Comment (roberlander2) [reply driven/storage/adapter_v3.go]: Thanks @mdryankov, I agree with you about not manipulating the filter object by adding new group IDs, so I think the way it is written now is fine.

File: driven/storage/adapter_v3.go
Comment (roberlander2) [reply driven/storage/adapter_v3.go]: Thanks @mdryankov, I think removing this block is fine. I have an additional comment on `FindGroupsV3` in general, since I believe it can be simplified to avoid some confusion and unnecessary DB requests.

I think the following can replace the 4 `if` blocks before the first `if` block modifying the filter. It looks to me that these 4 `if` blocks are used to find memberships and add their respective group IDs to `groupIDs`, so I think this can all be done with one DB request. Please let me know if I am missing something here or if you have any additional thoughts.
```
if filter.MemberUserID != nil || filter.MemberID != nil || filter.MemberExternalID != nil {
        // find group memberships
	memberships, err = sa.FindGroupMemberships(clientID, model.MembershipFilter{
		ID: filter.MemberID,
		UserID: filter.MemberUserID,
		ExternalID: filter.MemberExternalID,
	})
	if err != nil {
		return nil, err
	}

	for _, membership := range memberships.Items {
		if len(groupIDMap) == 0 || !groupIDMap[membership.GroupID] {
			groupIDs = append(groupIDs, membership.GroupID)
			groupIDMap[membership.GroupID] = true
		}
	}
}
```

Comment (mdryankov) [reply driven/storage/adapter_v3.go]: Yes. Thank you @roberlander2

File: driven/storage/adapter.go
Comment (mdryankov) [inline driven/storage/adapter.go:266]: @shurwit this artificial split of one and the same user unique per app is not good and it produces overhead in design and support. This is clear example of the misconception between the UIUC client and the admin client and the shared groups data (as a resource). The app authorisation should be addressed in different than the current way.

Comment (mdryankov) [reply driven/storage/adapter.go:266]: Nevermind. I had a talk with @petyos and he explained the situation and the design. I will remove the admin login API to stop the cross membership migration

**Pull request merged**
